examples/Protobuf/Program.cs

```diff
@@ -65,12 +40,14 @@ static async Task Main(string[] args)
                 AutoOffsetReset = AutoOffsetReset.Latest
             };
 
+            var messageParser = new MessageParser<User>(() => new User());
+
             var consumeTask = Task.Run(() =>
             {
                 // consume a single message then exit.
                 using (var consumer =
                     new ConsumerBuilder<int, User>(consumerConfig)
-                        .SetValueDeserializer(new ProtobufDeserializer<User>())
+                        .SetValueDeserializer((data, isNull) => messageParser.ParseFrom(data.ToArray()))

```

Comment:
What if the deserializer needs something else from the message, i.e., the value schema is in a message header, how would that look?

Reply:
the use-case you mention is something it'd be great to cater for, though note that it isn't actually covered by the `SerializationContext` struct as it currently stands. there are two ways the API could be extended to support this. 1. extend the `SerializationContext` struct. 2. Add an additional deserializer type.

the motivation for this PR is that the simple case is going to be very commonly needed (the most commonly needed), and common things should be easy, so a separate type is justified. The use case you bring up can be just as easily catered for by adding a the more comprehensive delegate type later (+ likely extending `SerializationContext`). Alternatively, we could require people use `IAsyncDeserializer` for this use case (not as good - i don't see a good reason for limiting the number of deserializer types, if they serve a purpose).